src/app-layer-parser.c

```diff
@@ -597,12 +597,12 @@ uint64_t AppLayerParserGetTransactionLogId(AppLayerParserState *pstate)
     SCReturnCT((pstate == NULL) ? 0 : pstate->log_id, "uint64_t");
 }
 
-void AppLayerParserSetTransactionLogId(AppLayerParserState *pstate)
+void AppLayerParserSetTransactionLogId(AppLayerParserState *pstate, uint64_t tx_id)
 {
     SCEnter();
 
     if (pstate != NULL)
-        pstate->log_id++;
+        pstate->log_id = tx_id;
```

Comment:
Yes, if thats the only TX.  But what it really does it update the LogId when all the transactions have been logged.  If some, but not all were logged it doesn't get updated, so the next loop starts earlier in the transaction list looking for tx's that are better to be logged.

I'm realizing this isn't very good.  So you have a constant stream of overlapping transactions, you may hit a state where there is always a tx behind the latest that is not ready to be logged, and this number will never get updated.

Instead I think we need to track the maximum tx id where is, and all transactions before it were completely logged, and update the LogId to that number. That will prevent the condition I describe above.